packages/controls/src/serialization/function-serialization.ts

```diff
@@ -44,30 +53,46 @@ export function serializeFunction<T extends AnyFunction>(
   return messageChannel.port2 as SerializedFunction<T>
 }
 
-export function deserializeFunction<T extends AnyFunction>(
-  serializedFunction: SerializedFunction<T>,
-): DeserializedFunction<T> {
-  type Result = SerializedFunctionReturnType<T>
-  type ResultMessageEvent = MessageEvent<[CallID, Result]>
-  type ResolveCallPromise = (value: Result | PromiseLike<Result>) => void
-
+function onmessageHandler<T extends AnyFunction>(): OnMessageHandler<T> {
+  type ResultMessageEvent = MessageEvent<
+    [CallID, SerializedFunctionReturnType<T>]
+  >
   let nextCallId = 0
-  const calls = new Map<CallID, ResolveCallPromise>()
+  const calls = new Map<CallID, ResolveCallPromise<T>>()
 
-  serializedFunction.onmessage = ({
+  const result: OnMessageHandler<T> = ({
     data: [callId, result],
   }: ResultMessageEvent) => {
     calls.get(callId)?.(result)
-
     calls.delete(callId)
   }
 
-  return function deserializedFunction(...args) {
-    return new Promise((resolve) => {
-      const callId = nextCallId++
+  result.newCall = (resolve: ResolveCallPromise<T>) => {
+    const callId = nextCallId++
+    calls.set(callId, resolve)
+    return callId
+  }
 
-      calls.set(callId, resolve)
+  return result
```

Comment:
Use a separate closure to implicitly attach the calls map and the call counter to the `onmessage` handler. The `newCall` method allows us to append to this state from the outside.

I originally tried to attach this state directly to the `MessagePort` object, but because it's a built-in type, it cannot be extended that way.